opentelemetry-sdk/src/metrics/mod.rs

```diff
@@ -245,6 +286,14 @@ mod tests {
         counter_aggregation_attribute_order_helper(Temporality::Cumulative, false);
     }
 
+    #[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+    async fn counter_aggregation_duplicate_attribute_keys() {
```

Comment:
This test doesn't test actual issue (it succeeds on `main` every single time, even though `main` has a bug).

I have wrote a test that actually fails on main everytime* (has 1% chance to succeed, due to Hasher being unpredictable), but succeeds on your PR. 
So your fix works :)

Here's the tests content.

```rust
    #[tokio::test(flavor = "multi_thread", worker_threads = 1)]
    async fn aggregation_attribute_keys_are_deduplicated_and_sorted() {
        // Run this test with stdout enabled to see output.
        // cargo test counter_aggregation_duplicate_attribute_keys --features=testing -- --nocapture

        // attributes are stored in HashMap, but default Hasher has unpredictable order everytime
        // to reduce test flakiness create many combinations
        for combination in 0..10 {
            aggregation_attribute_keys_are_deduplicated_and_sorted_helper(combination, Temporality::Delta);
            aggregation_attribute_keys_are_deduplicated_and_sorted_helper(
                combination,
                Temporality::Cumulative,
            );
        }
    }

    fn aggregation_attribute_keys_are_deduplicated_and_sorted_helper(
        combination: i32,
        temporality: Temporality,
    ) {
        // Arrange
        let mut test_context = TestContext::new(temporality);
        let counter = test_context.u64_counter("test", "my_counter", None);

        // duplicate zero(0)
        let keys = vec![0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
        let mut rng = thread_rng();

        // create 10 measurement with shuffled keys
        for _ in 0..10 {
            let mut shuffled = keys.clone();
            shuffled.shuffle(&mut rng);
            let shuffled: Vec<_> = shuffled
                .into_iter()
                .scan(false, |zero_seen, value| {
                    // make sure that for duplicated key (value=0) first time it will be 100, last value will be 0
                    Some(KeyValue::new(
                        format!("k{combination}.{value}"),
                        if value == 0 {
                            if *zero_seen {
                                // collector should always return last value (0)
                                value
                            } else {
                                // for zero value, if we see it for the first time, set it to 100,
                                // last value will always be 0,
                                *zero_seen = true;
                                100
                            }
                        } else {
                            value
                        },
                    ))
                })
                .collect();
            counter.add(1, &shuffled);
        }

        test_context.flush_metrics();

        // Assert
        let sum = test_context.get_aggregation::<data::Sum<u64>>("my_counter", None);
        assert_eq!(sum.data_points.len(), 1);
        
        let expected: Vec<_> = (0..10)
            .into_iter()
            .map(|v| KeyValue::new(format!("k{combination}.{v}"), v))
            .collect();
        let dp = sum.data_points.iter().next().unwrap();
        assert_eq!(dp.value, 10);
        assert_eq!(dp.attributes, expected);
    }
```

Reply:
> This test doesn't test actual issue (it succeeds on main every single time, even though main has a bug).

The test does fail on main. I suspect that you ran the test without updating `fn find_datapoint_with_key_value<'a, T>`.